### DISCUSSION TITLE
Why not use X (Ethereum, Cardano, Polkadot, etc)

### DISCUSSION THREAD

**Discussion by GGAlanSmithee** [feldmanchain/whitepaper#1] [General]
#  Why not use X (Ethereum, Cardano, Polkadot etc)

The technological scope of this project can most likely be implemented within most of the general-purpose blockchains that are available today. Why should we build our own, or shouldn't we?

Let's use this discussion to talk about pros and cons in regards to using existing tech, and if so, which one could be the most suitable for what we are trying to accomplish.

**Comment by GGAlanSmithee** [comment]
Let me preface this discussion with the fact that I would enjoy the challange, learning experience and technological freedom that building our own blockchain would come with. It would be extra awesome to be able to design a novel consensus mechanism, such as "Proof of Build".

However, for the good of the project, and with respect to available time, I think it would be good to keep an open mind towards using existing tech.

I am currently reading up on some of the larger "level one" blockchains to see how good of a fit they would be. This is somewhat hard since we haven't really formulated exactly what our needs are, so I think that would actually be a better place to start. I think I have a general idea of what that could be, and will try to start pinning it down in a not to distant future.

Even if we do not end up using an existing blockchain, reading up on them will certainly help inform us in our decisionmaking. Meanwhile, if anyone have any meaningful input or experience in regards to this choice, feel free to use this discussion as a place to share them, it will be of great help for us.

**Reply by brokenprogrammer** [reply]
I would like to join the discussion first with stating that I do not know much about different blockchains, I have heard about smart contracts and bitcoin's proof of work and understand them at a very basic level.

Before I give my oppinion I would like to read up a bit on different blockchains as well, I have composed a list of a few whitepapers that I have not yet read but will read to have something to back up my thoughts with.

Here is a list with different whitepapers from different technologies I picked somewhat random. I think all of the ones are trying to solve something different which will make the reads more interesting.

1. Bitcoin Whitepaper:  https://bitcoinwhitepaper.co/
2. Ethereum Whitepaper: https://ethereum.org/en/whitepaper/
3. Chainlink Whitepaper: https://research.chain.link/whitepaper-v2.pdf
4. TrueUSD Whitepaper: https://trueusd.com/pdf/TUSD_WhitePaper.pdf
5. Cosmos Whitepaper: https://github.com/cosmos/cosmos/blob/master/WHITEPAPER.md
6. Ripple Whitepaper: https://ripple.com/files/ripple_consensus_whitepaper.pdf
7. Monero Whitepaper: https://github.com/monero-project/research-lab/blob/master/whitepaper/whitepaper.pdf

I will get back to this discussion in some time with what I've learned.

**Reply by GGAlanSmithee** [reply]
Thanks for your input @brokenprogrammer . Feel free to share any research you might do in the repo. I myself will be adding md files in a `/research` folder, ie. `/research/ethereum.md`. Some of the above-mentioned technologies are rather specific in the problem they are trying to solve, and not all of them are blockchains (like chainlink, which is an oracle), but will surely be interesting and educational to read nonetheless.

**Reply by GGAlanSmithee** [reply]
There are some blockchains that implements another consensus mechanism called "Proof-of-Storage", such as [Filecoin](https://spec.filecoin.io/algorithms/pos/), that can be of interest to us if we want to solve the issue of storing the completed builds within the blockchain. That could be two separate issues though, and maybe something we want to consider at a later time.

**Reply by GGAlanSmithee** [reply]
It would be interesting to see if we can find a blockchain that shares the same characteristics/needs as the feldmanchain. For example, does it exist a blockchain that manages to keep the fees either fixed or predictable while still incentivizing participants to process transactions? For someone to rely on a (automated) build network, costs would have to be somewhat predictable and reasonable.

It is probably possible to implement that in the network itself, but would have to think about how that would be done. A "easy way out" of this potential dilemma would be to use a stable coin, but then the builders (possibly) pays an oppurtunity cost and loose out on potential upsides of a token gaining traction. Also, the legality of stable coins is in question.

**Reply by brokenprogrammer** [reply]
I think we need to somehow make it predictable, a job can vary in how "hard" it is to go through with depending on the machine the builders are running on. I don't like the idea of a stablecoin implementation but instead that the price is based of something that has to do with the task or some metric on your source code for example lines of code. Also important that this doesnt end up like ethereum and their gas prices.

**Reply by GGAlanSmithee** [reply]
@brokenprogrammer could not agree more. As we talked about in person, there is a lot of ideas floating around. To recap that particular conversion, one idea is to implement something like a closed-envelop auction:

1. A user puts a build request up for auction
    - The user can at this point put a max price limit
2. A auctioneer facilitates the auction
    - The auction is closed-envelop to prevent gaming the system (this might be unnecesary or even counterproductive)
    - Find the `n` lowest bids
    - Close the auction, send build requests to the workers with the winning bids
3. `n` builders completes the build
    - respond with the completed build
4. A verifier node (probably the same as the auctioneer) verifies the correctness of the builds
    - verify that the ouput checksums match and are correct
    - facilitate transacting the reward, distributed/split over the winning nodes

Architecurly the user, auctioneer and verifyer could be one and the same system, but would not have to be. Anyhow you draw the lines between the different sections of this model , it has the potential to take care of both the security aspect of the build not being tampered with as well as implementing a self-balancing and self-governing within the system itself.

**Comment by GGAlanSmithee** [comment]
@brokenprogrammer In regards to protecting the source code of the user, on top of other anti-theft counter-measures, do you think it will be enough to send the files over TSL / SSH to prevent man-in-the-middle and similar attacks? (ie. by monitoring the network, taking the files as they come in)

**Reply by GGAlanSmithee** [reply]
In my mind it _should_ be enough since that is what's used in similar scenarios in other places. I guess the fact that the builder nodes will be installed on someone elses computer makes me question everything :)

**Reply by brokenprogrammer** [reply]
In my mind most projects would just run git clone / pull from inside the vm and git uses SSL so I believe we should be set in that regard.. However I am not that good with security unfortunately, perhaps we should consult someone who knows this stuff

**Reply by GGAlanSmithee** [reply]
@brokenprogrammer sounds like a plan. I have a basic understanding of these technologies, but not more than that, so reaching out to someone more knowledgeable is a fine idea. From what I do know, I think it should be enough from a security perspective. TLS uses one-time symmetric keys for the actual transfer, which is fine, the thing that I am a bit unsure about is that I think setting up TLS at either point requires the issuance of a certificate, and I am not sure how that would fit into the system as a whole. That part, to mee, _feels_ like a traditional server-client relationship, with a trusted party, which is something we want to avoid IMO.

**Comment by GGAlanSmithee** [comment]
I have been thinking a bit about how/if the usage of a blockchain fits into all of this, and I actually do think it does. Since we are thinking about implementing something along the lines of an auctioneer pattern (see image at bottom of post), it implies that transactions will take place. These transactions cannot be handled by a trusted third-party, hence we face all of the same issues that blockchains aim to solve.

To handle the transactions we will need something along the lines of a smart contract that;

1. receives a build request
2. auctions it out to builders
3. accepts bids
4. picks winners and relays the actual buildrequest
5. receives the finished builds and verifies the correctness of them
6. executes the monitary transaction (be it a token or w/e)
7. sends the finished build to the user

Disregarding point 4 and 7, which is the sending of the actual files and finished build (and could be done off-chain), the remaining flow is something that could be easily coded as a smart contract on say Ethereum. It might actually make sense to do it like that because of the huge network effect that would bring.

If we do not implement this on an exsting blockchain, we should aim for compatability with the ERC20 standard to leave the door open for future cross-chain interoperability.

![auctioneer flow](https://github.com/feldmanchain/whitepaper/blob/main/assets/feldmanchain-auctioneer-flow.png?raw=true)

**Reply by GGAlanSmithee** [reply]
The one thing we would have to think hard about is how to incetivice users to actively use the system for what it is intended for. If the underlying asset is expected to appreciate over time, as popular crypto currencies tend to do, that incentives participants to HODL. Also, the extreme volatility and four year cyclic trends of more or less all exising cryptos makes it a bit hard to balance this in a near-term (probably many years / cycles) perspective.

**Comment by GGAlanSmithee** [comment]
# Feldmanchain, the decentralized application

I have been thinking **_a lot_** about this lately, and as I see it right now, I think the best course of action is probably for us to build the feldmanchain as a decentralized app (dapp) on an existing blockchain, rather than as its own blockchain.

## The Problem

As I am thinking about the different attack vectors of the current design WIP, it becomes clear to me that the underlying problem - and solution - is the decentralization of the system. This is an issue that we _can_ solve, but the more important question is, do we have to? Or do we even _want_ to? I think the answer to those questions are no.

## Rationale

Starting with the builders, it became clear that we can never _really_ trust one single builder to carry out a build in good faith, and to counteract this, we proposed to [introduce build redundancy and distribution](https://github.com/feldmanchain/whitepaper/discussions/1#discussioncomment-1562544). I actually think this idea is sound in its simpleness, and if that was the end of it, that would be great, but unfortunately it is not.

After solving the issue with the builders, the problem is then moved to the auctioneers. How can we select auctioneers in a trustless manner in a decentralized system? The TLDR is that we can, using a similar system as for the builders, or simply disincentivizing it by making it financial illogical to do so. However, as it turns out, this issue propagates to the clients, and then turns into the more general problem of securing a blockchain, AKA implementing a consensus mechanism like Proof of Work/Stake.

Now _this_ problem is an interesting one, but one that has been solved over and over again. We can definitely do this, I mean we can just clone one of the OS blockchains as a base if we wanted to, but this would just be an means to an end in our case. We would be doing much work only to facilitate decentralized project builds. It would also require feldmanchain to gain enough traction to be truly decentralized and secure, which is hard in itself.

## Solution

It is becoming more apparent that what we are building is not a decentralized ledger (AKA a blockchain), but a decentralized app, which lives on a blockchain. As it turns out, if we separate the issue of securing builds from the issues of securing transactions - everything related to facilitating builds - then we are left only with the original problem of securing builds, to which we have a proposed solution.

In doing so, we probably benefit greatly from the network effect of the hosting blockchain. There are still issues to be solved, such as fees and performance, but there are many options available that should suffice. We also need to think on how we would facilitate the fledmanchain as an on-premise solution, but I think reasoning about that without taking trustlessness into account is much more reasonable and easy.

thoughts @brokenprogrammer ?

**Reply by brokenprogrammer** [reply]
Does the trust really have to be built into the system can't the trust be a value that acts almost how AIs work with how many % sure they are its the correct answer?

Perhaps even building in AI in the system to evaluate how likely it is that the build was a sucessfull one. And / or putting that weight on the users to evaluate the numbers of how many times the project was built. Perhaps they want to re-run it if it was only built once or perhaps they dont care

**Reply by GGAlanSmithee** [reply]
Thanks for the reply.

That is an interesting idea actually. It is really easy to limit oneself when thinking hard on specific issues, so broadening the space of potential solutions is needed. I am not well versed within the field of AI, so I cannot speak to the feasibility of that particular solution, but I would love to look into it.

In regards to having a correctness score or something to that affect, that is also an interesting option. If we did that, we would simply put the burden on the client, wether or not to accept the build. We could broaden this to do something like BitTorrent where clients receives different scores depending on their performance, AKA if they went offline during uploads etc, which either makes it less likely for them to participate in the network, or outright banning them from doing so. Using social constructs like this is a super interesting alternative to comming up with a techinical solution, and would play well into a economic system based on a token.

To complement this thread with what we talked about offline:

To borrow more ideas from the BitTorrent protocol, we could use something like trackers / swarms for client discovery. This would make it super easy to participate in a sub set of the network (perhaps behind a VPN) for the on prem use-case. I actually prefer that to how Bitcoin [handles client discovery in cold start scenarios](https://github.com/feldmanchain/bitcoin/blob/master/src/chainparams.cpp#L121).

There are still points to iron out, such as how to handle the fair seeding of the network, and as always, the devil is in the details, but this sounds like a path forward.

**Reply by brokenprogrammer** [reply]
What about some sort of rating system between client/builder where you rate your builder with positive / negative depending on if the build succeeded. Most auctioning sites use this kind of system. Perhaps add some insentive to actually give someone the correct rating that tokens are not distributed in case of a conflict and bad rated users gets sent to the bottom of the pool when selecting builders.

**Reply by GGAlanSmithee** [reply]
> What about some sort of rating system between client/builder where you rate your builder with positive / negative depending on if the build succeeded. Most auctioning sites use this kind of system. Perhaps add some insentive to actually give someone the correct rating that tokens are not distributed in case of a conflict and bad rated users gets sent to the bottom of the pool when selecting builders.

Yes, this makes a lot of sense. I've been thinking along these lines as well, and it would probably go a long way to build confidence in the system, but maybe not when it comes to client-auctioneer-builder coordinated attacks. But I very much want us to facilitate social metrics and balancing as a way to build confidence between the peers of the network.
